### PR TITLE
Test simple Plonk circuit on all curves

### PR DESCRIPTION
**Pull request by vesselinux**
Add unit tests for the simple quadratic residue circuit for all curves. Addresses issue https://github.com/clearmatics/libsnark/issues/103, except the Edwards and BN128 curves (issue #108).

### CODE REVIEW

File: libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#include <array>
+
+namespace libsnark
+{
+
+/// Implementation of a dummy transcript hasher interface (see
+/// transcript_hasher.hpp). It returns the number of the elemnts in
+/// the hash buffer as an Fr element. Specialized over the curve
+/// field. See also class bls12_381_test_vector_transcript_hasher,
+/// which is specific to the BLS12_381 curve.
```
Comment (dtebbs) [inline libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp]: Doesn't seem to be returning the number of elements, but rather the number of bytes.

File: libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.tcc [hunk 2]
```diff
@@ -0,0 +1,99 @@
+/** @file
+ *****************************************************************************
+ * @author     This file is part of libff, developed by Clearmatics Ltd
+ *             (originally developed by SCIPR Lab) and contributors
+ *             (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+
+#ifndef LIBSNARK_ZK_PROOF_SYSTEMS_PLONK_TESTS_DUMMY_TRANSCRIPT_HASHER_CPP_
```
Comment (dtebbs) [inline libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.tcc]: CPP -> TCC

File: libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    // of elements in the buffer as an Fr value for the purposes of
+    // unit testing. TODO: to be replaced by a call to a proper hash
+    // function e.g. SHA2, BLAKE, etc.
+    libff::Fr<ppT> get_hash();
+
+    // clear the buffer (for now only for testing)
+    void buffer_clear();
+
+    // get buffer size
+    size_t buffer_size();
```
Comment (dtebbs) [inline libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp]: Seems unused.  Let's remove.

File: libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    void add_element(const libff::G2<ppT> &element);
+
+    // Dummy implementation of get_hash that simply returns the number
+    // of elements in the buffer as an Fr value for the purposes of
+    // unit testing. TODO: to be replaced by a call to a proper hash
+    // function e.g. SHA2, BLAKE, etc.
+    libff::Fr<ppT> get_hash();
+
+    // clear the buffer (for now only for testing)
+    void buffer_clear();
```
Comment (dtebbs) [inline libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp]: The concept of "buffer" is an implementation detail of specific hashers.  This seems only to be used to reset the hasher to allow it to be reused, and to produce invalid proofs.

Let's rename to something geenric like `reset`.

**Review by dtebbs — CHANGES_REQUESTED** [review]
Just a few tweaks.  Otherwise LGTM.

File: libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp (hunk 1 above)
Comment (vesselinux) [reply libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp]: Good point. Fixed comments to read _"returns the number of **bytes** in the hash buffer"_.

File: libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.tcc (hunk 2 above)
Comment (vesselinux) [reply libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.tcc]: Thanks. Fixed.

File: libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp (hunk 3 above)
Comment (vesselinux) [reply libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp]: Removed. Also from `bls12_381_test_vector_transcript_hasher` class .

File: libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp (hunk 4 above)
Comment (vesselinux) [reply libsnark/zk_proof_systems/plonk/tests/dummy_transcript_hasher.hpp]: Renamed. Also in `bls12_381_test_vector_transcript_hasher` class .

**Comment by dtebbs** [comment]
@vesselinux  I think this needs rebasing onto the destination branch.

**Review by dtebbs — APPROVED** [review]
LGTM after rebase.

**Comment by vesselinux** [comment]
> LGTM after rebase.

Hmm, I am running `git log --graph origin/plonk origin/103-plonk-simple-circuit-all-curves` and it looks like the branch is already rebased on `plonk`.

**Pull request merged**